### PR TITLE
snap-bootstrap: implement "run" mode in snap-bootstrap initramfs-mounts

### PR DESCRIPTION
**Pull request by mvo5**
This commit implements `snap-boostrap initramfs-mounts` run-mode.
With that the system should be able to boot normally and have the
right base/kernel available during the boot process.

boot: add boot.Modeenv.Kernel support

For the grub bootloader on the ubuntu-boot partition we want to
keep the kernel commandline as static as possible to avoid TPM
reseal operations for each kernel update.

This means that we will have a kernel with a fixed name on the
ubuntu-boot partition. But the initramfs needs information what
kernel snap we expect so that it can mount the kernel snap
and make the kernel modules available early in initramfs.

This commit adds "kernel" to the modeenv so that snapd can
set what kernel snap is expected via the modeenv mechanism
that is already used to set what base to use. The snap-bootstrap
initramfs-mounts code will have to do a cross-check to ensure
that the booted kernel matches the expected kernel and deal
with kernel reverts from the bootloader. This will happen in
subsequent commits.

### CODE REVIEW

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 1]
```diff
@@ -184,7 +185,42 @@ func generateMountsModeRecover() error {
...
+			return nil
+		}
+	}
+	// 2. read modeenv
+	modeEnv, err := boot.ReadModeenv(filepath.Join(dataDir, "system-data"))
+	if err != nil {
+		return err
+	}
+	// 3. mount base
+	isMounted, err := osutilIsMounted(filepath.Join(runMnt, "base"))
```
Comment (cmatsuoka) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go]: It seems that `err` is not really being used here?

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 2]
```diff
@@ -184,7 +185,42 @@ func generateMountsModeRecover() error {
...
+	if !isMounted {
+		base := filepath.Join(dataDir, "system-data", dirs.SnapBlobDir, modeEnv.Base)
+		fmt.Fprintf(stdout, "%s %s\n", base, filepath.Join(runMnt, "base"))
+		return nil
+	}
+	// 4. mount kernel
+	isMounted, err = osutilIsMounted(filepath.Join(runMnt, "kernel"))
```
Comment (cmatsuoka) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Same here.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go [hunk 3]
```diff
@@ -184,7 +185,42 @@ func generateMountsModeRecover() error {
...
+	dataDir := filepath.Join(runMnt, "ubuntu-data")
+
+	// 1. always ensure basic partitions are mounted
+	for _, d := range []string{seedDir, bootDir, dataDir} {
+		isMounted, err := osutilIsMounted(d)
+		if err != nil {
+			return err
+		}
+		if !isMounted {
+			fmt.Fprintf(stdout, "/dev/disk/by-label/%s %s\n", filepath.Base(d), d)
```
Comment (zyga) [inline cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Are we always safe against spaces here? Is there a need for a typical space-escape handling like various osutil mount-related parsing/generating functions?

Comment (mvo5) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Currently yes, the partition names are all very strictly defined, i.e. ubuntu-{seed,boot,data}. If we open this up later we could enforce no-spaces via the gadget validations.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 1 above)
Comment (mvo5) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Yes, silly me. Fixed.

File: cmd/snap-bootstrap/cmd_initramfs_mounts.go (hunk 2 above)
Comment (mvo5) [reply cmd/snap-bootstrap/cmd_initramfs_mounts.go]: Fixed here too, thank you :)

**Comment by xnox** [comment]
So, how will grub find out which kernel snap to loop mount and execute the kernel image and the initrd? By reading modenv too?

**Pull request merged**